backend/hidrive/helpers.go

```diff
@@ -0,0 +1,902 @@
+)
+
+// MemberType represents the possible types of entries a directory can contain.
+type MemberType string
+
+// possible values for MemberType
+const (
+	AllMembers       MemberType = "all"
+	NoMembers        MemberType = "none"
+	DirectoryMembers MemberType = api.HiDriveObjectTypeDirectory
+	FileMembers      MemberType = api.HiDriveObjectTypeFile
+	SymlinkMembers   MemberType = api.HiDriveObjectTypeSymlink
+)
+
+// SortByField represents possible fields to sort entries of a directory by.
+type SortByField string
+
+// possible values for SortByField
+const (
+	descendingSort             string      = "-"
+	SortByName                 SortByField = "name"
+	SortByModTime              SortByField = "mtime"
+	SortByObjectType           SortByField = "type"
+	SortBySize                 SortByField = "size"
+	SortByNameDescending       SortByField = SortByField(descendingSort) + SortByName
+	SortByModTimeDescending    SortByField = SortByField(descendingSort) + SortByModTime
+	SortByObjectTypeDescending SortByField = SortByField(descendingSort) + SortByObjectType
+	SortBySizeDescending       SortByField = SortByField(descendingSort) + SortBySize
+)
+
+var (
+	// Unsorted disables sorting and can therefore not be combined with other values.
+	Unsorted = []SortByField{"none"}
+	// DefaultSorted does not specify how to sort and
+	// therefore implies the default sort order.
+	DefaultSorted = []SortByField{}
+)
+
+// CopyOrMoveOperationType represents the possible types of copy- and move-operations.
+type CopyOrMoveOperationType int
+
+// possible values for CopyOrMoveOperationType
+const (
+	MoveOriginal CopyOrMoveOperationType = iota
+	CopyOriginal
+	CopyOriginalPreserveModTime
+)
+
+// OnExistAction represents possible actions the API should take,
+// when a request tries to create a path that already exists.
+type OnExistAction string
+
+// possible values for OnExistAction
+const (
+	// IgnoreOnExist instructs the API not to execute
+	// the request in case of a conflict, but to return an error.
+	IgnoreOnExist OnExistAction = "ignore"
+	// AutoNameOnExist instructs the API to automatically rename
+	// any conflicting request-objects.
+	AutoNameOnExist OnExistAction = "autoname"
+	// OverwriteOnExist instructs the API to overwrite any conflicting files.
+	// This can only be used, if the request operates on files directly.
+	// (For example when moving/copying a file.)
+	// For most requests this action will simply be ignored.
+	OverwriteOnExist OnExistAction = "overwrite"
+)
+
+// shouldRetry returns a boolean as to whether this resp and err deserve to be retried.
+// It tries to expire/invalidate the token, if necessary.
+// It returns the err as a convenience.
+func (f *Fs) shouldRetry(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	if fserrors.ContextError(ctx, &err) {
+		return false, err
+	}
+	if resp != nil && (resp.StatusCode == 401 || isHTTPError(err, 401)) && len(resp.Header["Www-Authenticate"]) > 0 {
```

Comment:
Hm, I'm not exactly sure what `this` is referring to. **TL;DR**: `TokenSource.Expire` was introduced to handle corruption of the access token in the config, or the revocation of tokens.

I'll explain how this code sprang into existence:
To test whether the backend fetches a fresh access-token automatically, I decided to manually insert an invalid token into the configuration by deleting one of the characters.
Unfortunately, this led to the token never being updated, because the `expiry` time of the token in the config still indicated it was valid. The error was not particularly helpful either:
```
2022/07/05 20:33:54 Failed to create file system for "TestHiDrive:": could not access root-prefix: Error "401": Unauthorized
```
I did not like this behavior, because it meant that if a user accidentally corrupted their access token, they would not be able to use rclone as long as the token is supposedly valid, and the error message is too unspecific to understand what went wrong.

So I thought I could use `TokenSource.Invalidate` to manually tell rclone to fetch a new token, if the backend decides the current one is definitely invalid. But this does not seem to be the use-case for `Invalidate`: It just sets the token to an empty string and does not update the configuration, so when issuing the next retry, `TokenSource.Token` simply reads the invalid token from the config again.
This is `TokenSource.Expire` was introduced, so backends could decide via custom logic when a token is definitely invalid and needs to be refreshed. We decided not to change the existing behavior of `oauthutil` and instead introduced this as a new method.

Besides tolerating corruption of the token in the config, this is also useful in situations where an access token can be revoked by the OAuth-sever, as this would lead to the same situation: rclone believes the token from the config to be valid, because the `expiry`-time was not reached yet, but the server denies the request.
With HiDrive specifically, access tokens are only revoked when their corresponding refresh token has been revoked, so a new token cannot be fetched anyways. But even in this case, at least with `TokenSource.Expire` the user is informed about the invalid refresh token as soon as the access tokens are revoked via a helpful error message, not just after the current access token was supposed to expire.